### PR TITLE
Live updates for chart widgets

### PR DESCRIPTION
**Pull request by andreasgebauer**
- support live updates for charts
- supply emit listeners with item as argument

### CODE REVIEW

File: web/app/widgets/chart/chart.widget.js [hunk 1]
```diff
@@ -183,6 +187,53 @@
...
+                                    });
+                                }
+                            });
+                        }
+                        dataset.push(data);
+
+                        // remove old values
+                        if(removeOutdatedValues) {
+                            var startDate = startTime();
+                            angular.forEach(vm.datasets, function(ds) {
```
Comment (kubawolanin) [inline web/app/widgets/chart/chart.widget.js]: How about reducing number of `forEach` here by combining it like so:
```javascript
angular.forEach(vm.datasets, function(ds) {
    if(fillValues && dataset !== ds) {
         ds.push({
             state: ds[ds.length-1].state,
             time: new Date()
         });
    }

    if(removeOutdatedValues) {
        var startDate = startTime();
        // values are sorted by date
        for(var i = 0; i < ds.length; i++) {
            if(ds[i].time > startDate) {
                ds.splice(0, i);
                break;
            }
        }
    }
});

dataset.push(data);
```

File: web/app/widgets/chart/chart.widget.js [hunk 2]
```diff
@@ -76,23 +76,27 @@
...
+                    case '2W': startDate.setTime(startDate.getTime() - 2*7*24*60*60*1000); break;
+                    case 'M': startDate.setTime(startDate.getTime() - 31*24*60*60*1000); break; //Well...
+                    case '2M': startDate.setTime(startDate.getTime() - 2*31*24*60*60*1000); break;
+                    case '4M': startDate.setTime(startDate.getTime() - 4*31*24*60*60*1000); break;
+                    case 'Y': startDate.setTime(startDate.getTime() - 12*31*24*60*60*1000); break;
+                    default: startDate.setTime(startDate.getTime() - 24*60*60*1000); break;
+                }
+                return startDate;
             }
+            var startDate = startTime();
```
Comment (kubawolanin) [inline web/app/widgets/chart/chart.widget.js]: Is it used anywher? I see same `startDate` variable in :216

**Comment by kubawolanin** [comment]
One more comment: please add your name to the [AUTHORS file](https://github.com/openhab/org.openhab.ui.habpanel/blob/master/AUTHORS) :-)
Awesome stuff!

File: web/app/widgets/chart/chart.widget.js (hunk 2 above)
Comment (andreasgebauer) [inline web/app/widgets/chart/chart.widget.js]: Yep, the start date is used initially when chart is initialized. When an update is received the timespan must be recalculated based on the current time.

**Comment by kubawolanin** [comment]
lgtm
@ghys could you take a look?

File: web/app/widgets/chart/chart.widget.js [hunk 3]
```diff
@@ -183,6 +187,56 @@
                     vm.interactiveChartOptions.series.push(seriesoptions);
                 }
 
+                // if set to true datasets which don't receive an update for a long time 
+                // will have their dataset filled with latest state when other datasets receive an update
+                var fillValues = true;
```
Comment (andreasgebauer) [inline web/app/widgets/chart/chart.widget.js]: Should it be possible to change the behaviour behind this switch by adding a configuration option to the config dialog?

File: web/app/widgets/chart/chart.widget.js [hunk 4]
```diff
@@ -183,6 +187,56 @@
...
+                // if set to true live updates are supported
+                var supportLiveUpdates = true;
```
Comment (andreasgebauer) [inline web/app/widgets/chart/chart.widget.js]: Should it be possible to change the behaviour behind this switch by adding a configuration option to the config dialog or should it be activate by default?

File: web/app/widgets/chart/chart.widget.js [hunk 5]
```diff
@@ -183,6 +187,56 @@
...
+                // removes outdated values (whoms timestamp is not within timespan) from the datasets
+                var removeOutdatedValues = true;
```
Comment (andreasgebauer) [inline web/app/widgets/chart/chart.widget.js]: Should it be possible to change the behaviour behind this switch by adding a configuration option to the config dialog or should it be activate by default?

**Comment by andreasgebauer** [comment]
just added a tab to the configuration dialog where the attributes can be set

**Comment by ghys** [comment]
So I'm AWOL for one week and @kubawolanin is already running things around here :smile: ... well thanks for your help.

@andreasgebauer I like it, it seems to be working nicely at first sight!
I have some concerns though about the "remove outdated values" option - why wouldn't you always want it? In other words, if it's working well, I think it should be the default, even not an option. If you keep adding values to the datasets without removing old ones, you're eating up memory that cannot be reclaimed, this will lead to problems if you stay on a dashboard with updating charts for a long time.

**Comment by andreasgebauer** [comment]
You're right. At least one could opt-out by forcing tthe removal of outdated values ;)

Just updated the code.

**Comment by ghys** [comment]
@andreasgebauer cool :) I will test it tonight.
In the meantime: the contributing rules require you to sign off all of your commits :)
I will squash them when merging anyway so if you could at least amend your last one with `git commit --amend -s`` and ``git push -f` - I will assume you sign off on all of them, with your approval.

**Comment by andreasgebauer** [comment]
Alright. Sorry about that. Should have read them completely :)

**Comment by ghys** [comment]
Perfect, thanks @andreasgebauer :)

**Pull request merged**